File: 0046-permutations/0046-permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function (nums) {
    function swap(i,j,nums){
        let temp = nums[i];
        nums[i]= nums[j];
        nums[j]= temp;
    }
    function solve(res, nums, idx) {
        if (idx >= nums.length) {
            const numscopy = [...nums]
            res.push(numscopy)
        }
        for (let i = idx; i <= nums.length - 1; i++) {
            swap(i, idx, nums);
            solve(res, nums, idx + 1);
            swap(i, idx, nums);


        }
    }


    const res = []

    let idx = 0;
    

    solve(res, nums, idx)

    return res

};

File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    void solve(vector<vector<int>>& result,  int index,
               vector<int>& nums) {
        // base case
        if (index >= nums.size()) {
            result.push_back(nums);
            return;
        }
        for (int i = index; i < nums.size(); i++) {
            swap(nums[i], nums[index]);
            solve(result, index + 1, nums);
            swap(nums[i], nums[index]);
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {

        int n = nums.size();
        int index = 0;
        vector<vector<int>> result;
        solve(result, index, nums);

        return result;
    }
};